packages/expo/cli/start/server/AsyncNgrok.ts

```diff
@@ -0,0 +1,205 @@
+
+  /** Info about the currently running instance of ngrok. */
+  private serverUrl: string | null = null;
+
+  constructor(private projectRoot: string, private port: number) {
+    this.resolver = new NgrokResolver(projectRoot);
+  }
+
+  /** Get the active pid from the running instance of ngrok. */
+  // TODO: Use this instead of a stored local value.
+  private getActivePid(): number | null {
+    return this.resolver.get()?.getActiveProcess?.()?.pid ?? null;
+  }
+
+  public getActiveUrl(): string | null {
+    return this.serverUrl;
+  }
+
+  /** Terminate the instance by the pid. */
+  private killInstance() {
+    const pid = this.getActivePid();
+    if (pid !== null) {
+      try {
+        process.kill(pid);
+      } catch (e) {
+        Log.error(`Failed to kill ngrok tunnel PID: ${pid}`);
+      }
+    }
+  }
+
+  /** Exposed for testing. */
+  async _getProjectHostnameAsync() {
+    // TODO: Maybe assert no robot users?
+    const username = getActorDisplayName(await getUserAsync());
+
+    return [
+      // TODO: Is this needed?
+      await this.getProjectRandomnessAsync(),
+      slugify(username),
+      // Use the port to distinguish between multiple tunnels (webpack, metro).
+      this.port,
+      NGROK_CONFIG.domain,
+    ].join('.');
+  }
+
+  async startAsync({ timeout }: { timeout?: number } = {}): Promise<void> {
+    // Ensure that force quitting can clean up ngrok.
+    installExitHooks(() => {
+      this.killInstance();
+    });
+
+    // Ensure the instance is loaded first, this can linger so we should run it before the timeout.
+    await this.resolver.resolveAsync({
+      // For now, prefer global install since the package has native code (harder to install) and doesn't change very often.
+      prefersGlobalInstall: true,
+    });
+
+    // Ensure ADB reverse is running.
+    if (!(await startAdbReverseAsync([this.port]))) {
+      // TODO: Better error message.
+      throw new CommandError(
+        'NGROK_ADB',
+        `Cannot start tunnel URL because \`adb reverse\` failed for the connected Android(s).`
+      );
+    }
+
+    this.serverUrl = await this._connectToNgrokAsync({ timeout });
+
+    Log.debug('[ngrok] Tunnel URL:', this.serverUrl);
+    Log.log('Tunnel ready.');
+  }
+
+  public async stopAsync(): Promise<void> {
+    Log.debug('[ngrok] Stopping Tunnel');
+
+    await this.resolver.get()?.kill?.();
+    this.serverUrl = null;
+  }
+
+  /** Exposed for testing. */
+  async _connectToNgrokAsync(
+    options: { timeout?: number } = {},
+    attempts: number = 0
+  ): Promise<string> {
+    // Attempt to stop any hanging processes, this increases the chances of a successful connection.
+    await this.stopAsync();
+
+    // Get the instance quietly or assert otherwise.
+    const instance = await this.resolver.resolveAsync({
+      shouldPrompt: false,
+      autoInstall: false,
+    });
+
+    let timer: NodeJS.Timeout | null = null;
+    const results = await Promise.race([
```

Comment:
If the timeout fires, there's still a connectToNgrokInternalAsync promise task running in the background -- it doesn't get canceled/aborted. This means when we retry, there could be two or three connectToNgrokInternalAsync calls running concurrently, which doesn't seem desirable.

Additionally, the ngrok library already has retry logic: https://github.com/expo/ngrok/blob/ae558edbabd9e116e7909aad12fe19188796dffc/index.js#L24-L39. Multiple levels of retries usually leads to confusion. We might need to have nested retries for now since there is no other way to change the "project randomness" but this code seems on shaky ground to me.